instrumentation/servlet/servlet-3/servlet-3.0/src/main/java/io/opentelemetry/instrumentation/auto/servlet/v3_0/Servlet30CountingHttpServletResponse.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.auto.servlet.v3_0;
+
+import io.opentelemetry.instrumentation.auto.servlet.v3.AbstractCountingHttpServletResponse;
+import java.io.IOException;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletResponse;
+
+public class Servlet30CountingHttpServletResponse extends AbstractCountingHttpServletResponse {
+  private CountingServletOutputStream outputStream = null;
+
+  /**
+   * Constructs a response adaptor wrapping the given response.
+   *
+   * @throws IllegalArgumentException if the response is null
+   */
+  public Servlet30CountingHttpServletResponse(HttpServletResponse response) {
+    super(response);
+  }
+
+  @Override
+  public ServletOutputStream getOutputStream() throws IOException {
+    if (outputStream == null) {
+      outputStream = new CountingServletOutputStream(super.getOutputStream());
+    }
+    return outputStream;
+  }
+
+  @Override
+  protected int getOutputStreamCounter() {
+    return outputStream == null ? 0 : outputStream.counter;
+  }
+
+  static class CountingServletOutputStream extends ServletOutputStream {
```

Comment:
Can this class be shared in servlet-common? It seems to only depend on servlet (technically only JDK)

Reply:
Actually this class is the reason why I had to split 3.0 and 3.1 instrumentations 😄 

3.1 introduced two new abstract methods and 1 new type reference (`WriteListener`) so I had to create 2 different `ServletOutputStream` implementations.